[feature] Add support to reversion to the REST API #894

dee077 opened the pull request:
## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

## Reference to Existing Issue

Closes #894.

## Description of Changes

Introduces three new REST API endpoints for managing revisions:

- **List all revisions** with optional filtering by model:  
`GET /controller/reversion/?model=<model_name>`


- **Inspect a specific revision using its ID**:  
`GET /controller/reversion/<revision_id>/`


- **Restore a revision based on its ID**:  
`POST /controller/reversion/<revision_id>/restore/`

Let me know if any changes are needed in the response data whether anything should be included, excluded, or modified. Also, please share any suggestions for adding more filters or adjusting the response format for the restore endpoint

dee077 committed:
[feature] Rivision for Rest Api changes  #894

Implemented three endpoints:
1. List all revisions with optional filtering by model.
2. Inspect a specific revision by its ID.
3. Restore a revision using its ID.

 Fixes #894

nemesifier commented on openwisp_controller/config/api/filters.py:
```diff
@@ -142,3 +143,18 @@ def __init__(self, *args, **kwargs):
+    model = filters.CharFilter(field_name="content_type__model", lookup_expr="iexact")
+
+    def _set_valid_filterform_labels(self):
+        self.filters["model"].label = _("Model")
```
We are using single quotes for strings, eg: `'model'`. **Please maintain consistency!**

nemesifier commented on openwisp_controller/config/api/urls.py:
```diff
@@ -83,6 +83,21 @@ def get_api_urls(api_views):
+                name='reversion_detail',
+            ),
+            path(
+                'controller/reversion/<str:pk>/restore/',
```
The URLs seem to be heavily focused on the `django-reversion` implementation rather than aligning with OpenWISP's approach.  

This was not the original intention of the issue. The goal is to introduce additional URLs for each resource (e.g., `Device`, `Template`, etc.) that supports reversion, allowing users to **list revisions, review revisions, and restore them**.  

For example, for `Template`:  
- `GET /controller/template/revisions/`  
- `GET /controller/template/revisions/{pk}/`  
- `POST /controller/template/revisions/{pk}/restore/`

nemesifier reviewed (changes requested):
Good work! The code is clean and well-structured.  

However, this PR seems to be missing its main goal: ensuring that every versioned model admin has a corresponding versioned REST API that logs revisions—tracking who changed what—each time a modification is made.  

Please check my comments below for further details.

dee077 commented:
I’ve added REST API revision support using django-reversion. Here’s a summary

### What’s Added

- `AutoRevisionMixin` uses RevisionMixin suggested by @pandafy which wraps API requests in reversion.create_revision().
- Automatically sets the user and a comment for each revision.

### New Endpoints:

- `controller/<str:model_slug>/revision/` – List revisions
- `controller/<str:model_slug>/revision/<str:pk>/` – Revision detail
- `controller/<str:model_slug>/revision/<str:pk>/restore/` – Restore a revision
These endpoints are placed above similar paths like `controller/device/<str:pk>/` to avoid conflicts, as pk will never be revision.

### How It Works:

- Each change creates an entry in the `reversion_revision` table.
- The corresponding object snapshot is stored in `reversion_version` with the same `revision_id`.

### Issues Noticed

- When using AutoRevisionMixin in the DeviceDetailView, the `devicegroup_change_handler` behaves unexpectedly, so the template doesn’t update on group change.
- Fails the test: test_device_api_change_group.
- Tried the mixin in other modules (connection, geo, pki), observed complexity increase (some entries show +10 additional SQL queries).

### Question
Given the unexpected behavior and added complexity:
Is this approach okay to proceed with, or alternative design for API revision support is needed?

nemesifier commented on openwisp_controller/config/api/urls.py:
```diff
@@ -13,6 +13,21 @@ def get_api_urls(api_views):
+                'controller/<str:model_slug>/revision/<str:pk>/restore/',
+                api_views.revision_restore,
+                name='revision_restore',
+            ),
```
What if we shorten this to just `model`? I think it would be clear enough and more concise. It's a small thing but nonetheless over time these simplifications help to keep maintenance sane.

nemesifier replied:
I have another doubt on this, how are we enforcing that we are enabling this only on the models that support reversions in this app? Eg: Device, Template, Vpn, etc. Would this work for any model?

dee077 replied:
Currently, empty revisions are being stored with only user and comment entries, which does not offer anything.

dee077 replied:
Added a fix for this

nemesifier commented on openwisp_controller/config/api/filters.py:
```diff
@@ -142,3 +143,18 @@ def __init__(self, *args, **kwargs):
+
+    class Meta:
+        model = Version
+        fields = ["model"]
```
Please use single quotes here and everywhere (we'll switch to double quotes later but we haven't done so yet so please maintain consistency).

nemesifier commented on openwisp_controller/config/api/serializers.py:
```diff
@@ -376,3 +377,34 @@ def update(self, instance, validated_data):
+    date_created = serializers.DateTimeField(
+        source="revision.date_created", read_only=True
+    )
+    comment = serializers.CharField(source="revision.comment", read_only=True)
```
use single quotes

nemesifier commented on openwisp_controller/config/api/serializers.py:
```diff
@@ -376,3 +377,34 @@ def update(self, instance, validated_data):
+            "object_repr",
+            "user_id",
+            "date_created",
+            "comment",
```
same here and below

nemesifier commented on openwisp_controller/mixins.py:
```diff
@@ -35,3 +38,14 @@ class RelatedDeviceProtectedAPIMixin(
+    def dispatch(self, request, *args, **kwargs):
+        if request.method in ('GET', 'HEAD', 'OPTIONS'):
+            return super().dispatch(request, *args, **kwargs)
+        with reversion.create_revision():
```
Wouldn't this trigger also on DELETE? May not be desired, since DELETE wouldn't have any change and would store an additional revision identical to the previous one.

I would do the opposite: trigger only on PUT/PATCH/POST. We need a test for each HTTP method.
POST may have to create the initial revision.

nemesifier reviewed (changes requested):
Looks on the right track.

I will do a manual round of testing soon, in the meantime see my comments below.

We'll need to add this to `pki`, `geo` and `connection`.

pandafy commented on openwisp_controller/mixins.py:
```diff
@@ -35,3 +38,14 @@ class RelatedDeviceProtectedAPIMixin(
     pass
+
+
+class AutoRevisionMixin(RevisionMixin):
+    def dispatch(self, request, *args, **kwargs):
+        if request.method in ('GET', 'HEAD', 'OPTIONS'):
+            return super().dispatch(request, *args, **kwargs)
+        with reversion.create_revision():
+            response = super().dispatch(request, *args, **kwargs)
+            reversion.set_user(request.user)
+            reversion.set_comment(f'API request: {request.method} {request.path}')
+            return response
```
@dee077 doesn't RevisionMixin already handles this? 

https://github.com/etianen/django-reversion/blob/9a7cd7419121e56d65247b050b7aa8a105aa000c/reversion/views.py#L36-L60

Apart from adding a comment that this change was made from the REST API, we are not doing any extra. In my opinion, the comment does not add much value.  While reviewing the list of revisions (at the time of restoring revisions). I would prefer simplifying the logic here and use the default behavior.

nemesifier replied:
@pandafy I think knowing the change was made via the API is useful for our commercial support service, if users are doing changes automatically via the REST API and something breaks due to those changes we can find out with less time, so I think this is a plus worth keeping.

pandafy commented on openwisp_controller/config/api/serializers.py:
```diff
@@ -376,3 +377,34 @@ def update(self, instance, validated_data):
+
+    class Meta:
+        model = Version
+        fields = [
+            "id",
+            "revision_id",
+            "object_id",
+            "content_type",
+            "db",
+            "format",
+            "serialized_data",
+            "object_repr",
+            "user_id",
+            "date_created",
```
Thinking out loud: 

Do we really need to expose all these details in the API? The admin views for listing versions don't show all these details. I think we can hide internal details (e.g. db, format, object_repr, content_type)

nemesifier replied:
I think that if it's not expensive in terms of processing (number of queries, CPU usage), we can return this info in the API.
It doesn't have to be identical to the admin, it's ok to have more technical details in the REST API.
What would not be allowable in my opinion would be the opposite: having the REST API show less details than the admin.

dee077 commented:
## Updates
- Fix single quotes
- Added the `AutoRevisionMixin` across the module. This caused 2 big issues
- The test was failing due to `AutoRevisionMixin` wrapping the view in a `transaction.atomic()` block, which delayed the `post_save signal` (registered via `transaction.on_commit`) until the outermost transaction committed. Since the outer transaction wasn’t triggered when PUT request generating the response,that returned outdated data while the database was updated afterward. Resolved by setting revision_atomic = False in the mixin to avoid unnecessary nesting.
- Some `GeoTestsApi` were inconsistently counting queries due to Django Reversion caching the `ContentType` lookup (used for versioning). On the first run, the lookup incurred extra queries, on subsequent runs, caching reduced them, leading to assertion failures. Fixed by adding `ContentType.objects.clear_cache()` in the `setUp method` of the test case to ensure consistent query counts.

nemesifier reviewed (commented):
Can you point to the lines of code that are generating the additional queries?

Have you found out what additional queries are being generated that before weren't?

If we can see the query it would be faster to verify whether all the queries are needed, the increase seems pretty significant.

pandafy commented on openwisp_controller/config/api/views.py:
```diff
@@ -289,6 +302,58 @@ def certificate_delete_invalidates_cache(cls, organization_id, common_name):
+class RevisionListView(ProtectedAPIMixin, AutoRevisionMixin, ListAPIView):
+    serializer_class = ReversionSerializer
+
+    def get_queryset(self):
+        model_slug = self.kwargs.get('model').lower()
+        return (
+            Version.objects.select_related('revision')
+            .filter(content_type__model=model_slug)
+            .order_by('-revision__date_created')
+        )
```
The view name contains Revision, but the queryset is returning `Version` objects. This is confusing. 

If I understand correctly, a revision is made up of multiple versions (from different objects). 

Is it logically possible to restore a particular version of the object while others are unaffected? Could this potentially render the database in a broken state?

pandafy commented on openwisp_controller/config/api/views.py:
```diff
@@ -289,6 +302,58 @@ def certificate_delete_invalidates_cache(cls, organization_id, common_name):
         cls.get_device_group.invalidate(cls, org_slug, common_name)
 
 
+class RevisionListView(ProtectedAPIMixin, AutoRevisionMixin, ListAPIView):
```
Why does this view needs `AutoRevisionMixin`?

pandafy commented on openwisp_controller/config/api/views.py:
```diff
@@ -289,6 +302,58 @@ def certificate_delete_invalidates_cache(cls, organization_id, common_name):
+    def post(self, request, *args, **kwargs):
+        qs = self.get_queryset()
+        versions = get_list_or_404(qs, revision_id=kwargs['pk'])
+        with reversion.create_revision():
+            for version in versions:
+                version.revert()
+            reversion.set_user(request.user)
+            reversion.set_comment(
+                f"Restored to previous revision: {self.kwargs.get('pk')}"
+            )
```
Shall we flag this operation as an atomic transaction?

pandafy reviewed (changes requested):
> Can you point to the lines of code that are generating the additional queries?
> Have you found out what additional queries are being generated that before weren't?

Here's a diff of the queries generated by the `test_patch_deviceconnectoin_detail`, 

https://www.diffchecker.com/ToUeQeyZ/

From my analysis, it appears that django-rivision creates a version for all related objects (here - DeviceConnection, Config and Device). This is inline with VersionAdmin behaviour.

dee077 commented:
## Updates
- Again added test with increased queries.
- The queryset uses Version objects because there is no reverse relationship from Revision to Version, but each Version is linked to a Revision.
- Rename `RevisionSerializer` to `VersionSerializer`
- In `VersionSerializer` directly defined the fields instead of using SerializerMethodField
- Added a filter by revision_id in RevisionListView to allow retrieving versions linked to the same revision, which helps group related changes made at the same time.
- Added test for revision_id filter
- Small qa fix for double quotes.

dee077 commented:
## Updates 
- Added additional check for registered model

With this approach, only registered models will store revisions, while unregistered models can still include the mixin without any side effects. This allows future developers to safely add the mixin to new POST, PUT, or PATCH requests without needing to worry about whether the model is registered with reversion.

nemesifier reviewed (changes requested):
@dee077 can you please run `openwisp-qa-format` here again with the latest version of openwisp-utils? See https://github.com/openwisp/openwisp-utils/pull/456.

pandafy reviewed (changes requested):
I did a round of manual testing with the `Templates` model. Changing the model fields from the API endpoints creates revisions as expected. Kudos! 

---

Is it possible to list all revisions of a particular object? Similar to this view 
https://demo.openwisp.io/admin/config/template/e450012e-a482-48ad-9b2b-f0faa93b9cac/history/

![Screenshot from 2025-06-10 18-42-38](https://github.com/user-attachments/assets/20d4f79a-b3c7-4f9f-b827-cb743a68efe4)

If not, it really hurts the functionality of the REST API endpoints. 

This end point is missing from the issue description. Let's discuss it with @nemesifier first before proceeding on this front.

dee077 committed:
[fix] Rebase

nemesifier reviewed (changes requested):
@dee077 there's conflicts with master

dee077 committed:
[fix] Update endpoints and tests

dee077 commented on openwisp_controller/connection/api/views.py:
```diff
@@ -131,15 +133,38 @@ def get_serializer_context(self):
+            assert self.get_parent_queryset().exists()
+        except (AssertionError, ValidationError):
+            device_id = self.kwargs["device_id"]
+            raise NotFound(detail=f'Device with ID "{device_id}" not found.')
```
improper rebase

dee077 commented:
### Question
- Should we use all these filters or reduce them
- Is the increase in no of queries too much? Should I spend some more time finding out if there is a way to reduce them?

### Pending
- Add more test scenarios based on the filter
- Update no of queries to fix failing tests

nemesifier reviewed (commented):
@dee077 why are the queries almost doubling?

dee077 committed:
[fix] Test and reduece some queries

dee077 commented:
> @dee077 why are the queries almost doubling?

Let's say if the `DeviceLocation` model is registered to reversion like this
```
reversion.register(
    model=DeviceLocation,
    follow=["device", "location"],
)
```
When DeviceLocation is updated, reversion also versions `DeviceLocation`, `Device`, and `Location` in the same revision. 

This triggers additional ContentType lookups, related object fetches, PK collection, serialization of each object, and finally one `reversion_revision` insert plus multiple `reversion_version` entries for all tracked objects in that revision, which is why the query count increases significantly.

dee077 commented on openwisp_controller/config/api/views.py:
```diff
@@ -289,6 +303,60 @@ def certificate_delete_invalidates_cache(cls, organization_id, common_name):
+        serializer = VersionSerializer(
+            versions, many=True, context=self.get_serializer_context()
+        )
+        return Response(serializer.data, status=status.HTTP_200_OK)
```
Updated!

dee077 commented on openwisp_controller/config/tests/test_api.py:
```diff
@@ -1634,3 +1635,49 @@ def test_device_patch_with_templates_of_same_org(self):
+            response = self.client.get(f"{path}?revision_id={revision_id}")
+            response_json = response.json()
+            self.assertEqual(response.status_code, 200)
+            self.assertEqual(len(response_json), 2)
```
Todo

dee077 committed:
[fix] Multi-tenancy for reversion endpoints

dee077 commented on openwisp_controller/connection/tests/test_api.py:
```diff
@@ -561,7 +561,7 @@ def test_post_deviceconnection_list(self):
             "failure_reason": "",
         }
-        with self.assertNumQueries(12):
+        with self.assertNumQueries(23):
```
Decided to go with the increase!
